### ISSUE TITLE
Create basic LinkedList visualization based on VisualizationApp

### ISSUE THREAD

**Issue by JMCanning78** [JMCanning78/datastructures-visualization#74]
Create a LinkedList class based on VisualizationApp using the code
from LinkedList.py.  The box and pointer visualization of nodes is
already done but we also need a visual representation of the
LinkedList object as well.  This will show the head of the linked list
that corresponds to the `LinkedList` object in the textbook code.  The
only thing in that object is a (possibly empty) pointer to the first
`Link` object (so maybe a simple "first:" label would do).

The last `Link` in the list should have just a dot for it's `next` pointer
instead of an arrow to indicate the `None` value stored there.

Allow links to store short text or numeric values, maybe 8 characters
max.  Place links in successive rows on the canvas, moving to a new
row whenever inserting a new link would go off or too near the edge
of the canvas to show the next pointer.

Implement the animated methods for:
 * New Linked List (empty)
 * Get First
 * Insert
 * Search
 * Delete

Clicking on a `Link` item should update the argument for the Search,
Insert and Delete operations to be that item's value.

**Comment by JMCanning78** [comment]
Here's the textbook code for linked lists:

```
class Link(object):            # One datum in a linked list
   def __init__(self, datum, next=None): # Constructor
      self.__data = datum      # The datum for this link
      self.__next = next       # Reference to next Link

   def getData(self):          # Return the datum stored in this link
      return self.__data

   def setData(self, datum):   # Change the datum in this Link
      self.__data = datum

   def getNext(self): return self.__next # Return the next link

   def setNext(self, link):    # Change the next link to a new Link
      if link is None or isinstance(link, Link): #Must be Link or None
         self.__next = link
      else:
         raise Exception("Next link must be Link or None")

   def isLast(self):           # Test if link is last in the chain
      return self.getNext() is None  # True iff no next Link

   def __str__(self):          # Make a string representation of link
      return str(self.getData())

class LinkedList(object):      # A linked list of data elements
   def __init__(self):         # Constructor
      self.__first = None      # Reference to first Link

   def getFirst(self): return self.__first # Return the first link

   def setFirst(self, link):   # Change the first link to a new Link
      if link is None or isinstance(link, Link): #Must be Link or None
         self.__first = link
      else:
         raise Exception("First link must be Link or None")
      
   def getNext(self): return self.getFirst()    # First link is next
   def setNext(self, link): self.setFirst(link) # First link is next

   def isEmpty(self):          # Test for empty list
      return self.getFirst() is None  # True iff no first Link

   def first(self):            # Return the first item in the list
      if self.isEmpty():       # as long as it is not empty
         raise Exception('No first item in empty list')
      return self.getFirst().getData() # Return data item (not Link)
   
   def traverse(self,          # Apply a function to all items in list
                func=print):   # with the default being to print
      link = self.getFirst()   # Start with first link
      while link is not None:  # Keep going until no more links
         func(link.getData())  # Apply the function to the item
         link = link.getNext() # Move on to next link

   def insertFirst(self, datum): # Insert a new datum at start of list
      link = Link(datum,       # Make a new Link for the datum
                  self.getFirst()) # What follows is the current list
      self.setFirst(link)      # Update list to include new Link

   def find(                   # Find the 1st Link whose key matches
         self, goal, key=identity): # the goal
      link = self.getFirst()   # Start at first link
      while link is not None:  # Search until the end of the list
         if key(link.getData()) == goal:  # Does this Link match?
            return link        # If so, return the Link itself
         link = link.getNext() # Else, continue on along list
         
   def search(                 # Find 1st item whose key matches goal
         self, goal, key=identity):
      link = self.find(goal, key) # Look for Link object that matches
      if link is not None:     # If found,
         return link.getData() # return its datum

   def insertAfter(            # Insert a new datum after the first
         self, goal, newDatum, # Link with a matching key
         key=identity):
      link = self.find(goal, key)  # Find matching Link object
      if link is None:         # If not found,
         return False          # return failure
      newLink = Link(          # Else build a new Link node with
         newDatum, link.getNext()) # new datum and remainder of list
      link.setNext(newLink)    # and insert after matching link
      return True

   def delete(self, goal,      # Delete the first Link from the
              key=identity):   # list whose key matches the goal
      if self.isEmpty():       # Empty list? Raise an exception
         raise Exception("Cannot delete from empty linked list")

      previous = self          # Link or LinkedList before Link
      while previous.getNext() is not None: # to be deleted
         link = previous.getNext()  # Next link after previous
         if goal == key(link.getData()): # If next Link matches,
            previous.setNext(  # change the previous' next
               link.getNext()) # to be Link's next and
            return True        # return since match was found
         previous = link       # Advance previous to next Link
         
      # Since loop ended without finding item, raise exception
      raise Exception("No item with matching key found in list")
```

The corresponding buttons in the visualzation and their "public" methods are:

Button        |    `LinkedList` method
----------- | ----------------------
 New    |  `__init__()`
 Get First | `first()`
 Insert | `insertFirst()`
 Search | `search()`
 Delete | `delete()`

The animation of `search()` involves the call to `find()` .  When we add code highlighting, those will be two different call environments.

**Comment by JMCanning78** [comment]
@EstherWerblowsky writes:

> 1. When I create a newList - the delete and getFirst buttons which don't take parameters are still enabled. Should I allow them to be enabled and just show an error message if they are pressed when unable to be used? or should I disable them until able to be used? -If the answer is yes to the second one, what is the best way of doing that in the code?
> 
> 2. What exactly do you mean when you write: "Clicking on a Link item should update the argument for the Search, Insert and Delete operations to be that item's value."?

For question 1, yes allow the Delete and Get First buttons to be enabled.  We have that same situation on other structures like stacks, queues, and arrays.  We've been letting users click them and get a message saying the structure is empty (or full or the item would go off the canvas, etc.).

For question 2, it's basically asking for the same behavior that is shown in Array.py and SimpleSorting.py.  If you click on one of the items on the canvas that represents a value in the array, the value is put into into the text entry widget in the Operations panel.  That enables searching or deleting for the next operation while ensuring a specific, existing key was entered.  In the case of the list, clicking on the data field of a `Link` object should copy that data into the argument box.  Clicking on the next field (the one with pointer) could also copy that same data, but I'm less comfortable with that. I guess it's OK since there's only one datum that's associated with `Link` and we aren't allowing any operations on pointers to `Link`'s.  The same situation arises with binary trees and skip lists. With 2-3-4 trees, however, there are multiple data in the node, and clicking anywhere other than an individual datum would be ambiguous.

**Comment by JMCanning78** [comment]
This was closed by PR #112
